### ISSUE TITLE
LogiX Displays randomly stop updating (freeze) over time.

### ISSUE THREAD

**Issue by KierDran** [Frooxius/NeosPublic#730]
Display nodes that should be constantly updating froze. When it was evident it was doing it, (in a world I hosted), I fully restarted Neos and went to a world that Electronus was hosting, and it happened there too. Keep in mind that Electronus never saw the issue, in either world, and I saw it in both.

**Comment by KierDran** [comment]
[Kier-LogiX Display Freezing.zip](https://github.com/Frooxius/NeosPublic/files/4861772/Kier-LogiX.Display.Freezing.zip)
![unknown](https://user-images.githubusercontent.com/13511918/86312235-7ef04c80-bbdf-11ea-9556-69e3fcdecb96.png)
(Logs and screenshot)

**Comment by Frooxius** [comment]
Hmm this is strange. Does any other text in-game update when this happens? What if you make another piece of LogiX that has something very simple, like Time (T) plugged into a display, will that update?

It sounds like either the background threads responsible for generating the graphics are clogged or there's some internal exception, logs would help a lot in that case.

**Comment by shiftyscales** [comment]
They provided logs, @Frooxius:
[Kier-LogiX_Display_Freezing.zip](https://github.com/Frooxius/NeosPublic/files/4861810/Kier-LogiX_Display_Freezing.zip)
```
System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.UIX.Text.Preprocess () [0x00047] in <fddbb27d1cce43d9a52839450d9f2d87>:0 
5:28:27 PM.736 ( 60 FPS)	Exception in Preprocess for text. String: Z: -78.60°
```

**Comment by shiftyscales** [comment]
They also specified "As for the logix... The only thing that matters is the start of it... the global xform, and the global to local. Then display it"

**Comment by Frooxius** [comment]
Thanks! Yeah this looks like a straighforward bug, should have a fix in the next build!

**Comment by Frooxius** [comment]
Hmm actually looks a bit more complex. Looks like it's potentially a race condition due to memory access issue. What CPU are you using?

I'm going to add some memory barriers in the next build, see if it helps once it comes out.

**Comment by KierDran** [comment]
I'm using a Intel i7-9700 @ 3GHz (And 48 GB of ram)

Btw, I did a test where I used a Random LogiX node (1 - 1000.1) and 16 displays off of it. Within 10 min, 9 of them "stopped".

**Comment by KierDran** [comment]
What? I didn't mean to close it!!!

**Comment by Electronus** [comment]
we did make a set up to replicate this issue. Just hooked up a random node to a bunch of display nodes. My side was completely fine for the duration of our test, and I think Kier had somewhere around 10 of them freeze. 
![2020-07-01 21 33 31](https://user-images.githubusercontent.com/64871170/86318229-9b47b580-bbee-11ea-804f-28e70b51e30d.jpg)

**Comment by Frooxius** [comment]
I see. It's really odd, because the CPU's should generally have strong memory ordering. It might be some implementation detail on Mono or something. It's technically nothing to do with the nodes themselves, just UIX text rendering, specifically multithreading it seems.

Normally this shouldn't happen due to strong memory ordering, but perhaps on your system this doesn't seem to work correctly? 

Once the next build comes out (I'll ping you in the release notes) give it a try and see if it still keeps happening.

**Comment by Hayden-da-Serg** [comment]
Could this potentially be related to UIX canvases not updating on occasion? I always get cases of all canvases not updating for a while (context menus, dash, inventory, Logix panel, literally any UIX canvas), sometimes lasts a few seconds sometimes lasts up to 5 minutes, I most notably see it when assets are loading in a form, either when loading a world or an avatar of sorts.
I personally have a Ryzen 7 3750h and 16GB of memory.
I haven't seen this particular issue yet, though I think it might be somewhat related.

**Comment by shiftyscales** [comment]
Oh? That sounds like it could potentially be one of the other scenarios @Frooxius had mentioned, @Hayden-da-Serg.

`It sounds like either the background threads responsible for generating the graphics are clogged or there's some internal exception, logs would help a lot in that case.`

**Comment by Frooxius** [comment]
This might be fixed in 2020.7.24.446, can you please check if it's still occurring?

**Comment by Frooxius** [comment]
Confirmed to be fixed in 2020.8.3.1234! Thanks to @KierDran for testing out instrumented builds to help narrow down the cause. Was an unfortunate race condition in UIX multithreading.
